### PR TITLE
Removing python shebangs and adding python3 virtual env to callsites

### PR DESCRIPTION
**Pull request by jmherbst**
This is a follow up to https://github.com/mcguirepr89/BirdNET-Pi/pull/287

There is not a great way to hardcode fully-qualified (no $HOME var) paths into shebangs for the python virtualenv so this is an attempt to force the correct python3 virtualenv use for all python script callsites.

For the `birdnet_stats.service` I tested that it runs fine without the shebang at all.  I think this is because the `streamlit` call is already using that virtualenv.

Sorry about the whitespace changes in the diff.  If you'd like I can remove those from my commits.  Or you can review with whitespace changes removed from diff with https://github.com/mcguirepr89/BirdNET-Pi/pull/288/files?diff=unified&w=1

### CODE REVIEW

**Comment by jmherbst** [comment]
Merging this will work for all new installs, but I'm not sure anything in a normal `update` will for the re-install of services, will it?

🤔

**Comment by mcguirepr89** [comment]
Sorry I'm moving slowly on this -- there are several plates spinning and I'm just trying to be careful while managing them

File: scripts/update_birdnet_snippets.sh [hunk 1]
```diff
@@ -14,6 +14,11 @@ if [ -f $my_dir/privacy_server ] || [ -L /usr/local/bin/privacy_server.py ];then
   rm -f $my_dir/privacy_server.py
   rm -f /usr/local/bin/privacy_server.py
 fi
+
+# Adds python virtual-env to the python systemd services
+sudo -E sed -i "s|ExecStart=.*|ExecStart=$HOME/BirdNET-Pi/birdnet/bin/python3 /usr/local/bin/server.py|" ~/BirdNET-Pi/templates/birdnet_server.service
```
Comment (mcguirepr89) [inline scripts/update_birdnet_snippets.sh:19]: Woah!! I didn't know `sed` could use the pipe as a delimiter like that! Super helpful for readability when paths are involved. :+1:

Comment (jmherbst) [reply scripts/update_birdnet_snippets.sh]: Yeah! I learned this because `$HOME` contained an un-escaped `/` and it was giving sed some issues.

**Comment by jmherbst** [comment]
I wonder if there'd be a good way to test using alternate remotes/branches? 🤔 💭

**Comment by mcguirepr89** [comment]
There HAS to be a way to effectively test updates while also being part of development :thinking: let's keep thinking :thinking: branching perhaps, but the most likely avenue will be learning the nitty gritty of `git` such that we can mimick an update after pushing a commit by checking out the commit prior to the push, or something nitty gritty like that :weary:

**Comment by mcguirepr89** [comment]
I like that approach -- I think it is a robust update method. Let's push that into this PR -- I'll check out this PR and do some mucking around and see what I can break. My hope is that I can checkout the PR, run the updater (after you push those changes) and it should just work. Then, I'll do a PR from the `ui-tweaks` branch into this fork. It might take a small bit of resolving to do, but once those changes have been pull into this fork, I want to try another update and see how it goes.

Does that make sense as a reasonable and worth-while test?

**Comment by jmherbst** [comment]
I think that sounds good - I pushed up a786e51.

One issue you may encounter is using this new update_birdnet.sh script to get a different branch may remove the changes to the script made in this branch's `a786e51`

**Comment by mcguirepr89** [comment]
We'll see how it goes -- if needed, other branches will have to merge the new updater ¯\_(ツ)_/¯ "crazy world, lotta updaters" -- Michael Gary Scott

**Comment by Nachtzuster** [comment]
I think testing upgrading might work like this (not tested):
->  on development machine
assuming origin points to you own github repo
assuming you are working on `cool_feature` branch
make sure your main is up to date with mcguirepr89/BirdNET-Pi
```
git switch main
git pull
git checkout -b test_cool_feature
git push --set-upstream origin test_cool_feature
```

-> on RPi
```
git remote add dev https://github.com/Nachtzuster/BirdNET-Pi.git
git fetch dev
git checkout remotes/dev/test_cool_feature
git pull dev test_cool_feature
```
 update on the ui to ensure you _are_ running main

-> on development machine: fake merging
```
git merge cool_feature
git push
```

-> on RPi
hit update in the gui to test

**Comment by mcguirepr89** [comment]
My steps so far, (they feel weird and wrong, but that's always my `git` reaction [pun intended]) starting on mcguirepr89/BirdNET-Pi.git main:
```
git checkout -f
git pull
```
```
git fetch origin pull/288/head:new_shebangs
git checkout -f new_shebangs
git remote add dev https://github.com/jmherbst/BirdNET-Pi.git   
git branch --set-upstream-to=dev/remove-python-shebangs new_shebangs
```


Now I wanted to test `update_birdnet.sh`, which without the flags, would put me back on my `main`.

So, I aimed to test `update_birdnet.sh` with the proper flags, but see the results below:

```
birdnetpi1:~/BirdNET-Pi$ update_birdnet.sh -r dev -b remove-python-shebangs
Error: remote 'dev' not found. Add the upstream remote to your repository and try again.
```
I needed to allow the output from `git remote show $remote` for some reason. You can see it fails if stderr is redirected to `/dev/null` (which is kind of strange that it would affect the exit code?) but succeeds if I remove `2&> /dev/null` from the script.
```
birdnetpi1:~/BirdNET-Pi$ bash <(sed "s|2&> /dev/null||g" scripts/update_birdnet.sh) -r dev -b remove-python-shebangs
* remote dev
  Fetch URL: https://github.com/jmherbst/BirdNET-Pi.git
  Push  URL: https://github.com/jmherbst/BirdNET-Pi.git
  HEAD branch: main
  Remote branches:
    Nachtzuster_improvements_to_birdnet_analysis tracked
    bashrc-removal                               tracked
    caddy-fix                                    tracked
    caddy-pwd-fix                                tracked
    change-python-shebangs                       tracked
    fix-python-shebangs                          tracked
    main                                         tracked
    mariadb                                      tracked
    notifications                                tracked
    python-lint                                  tracked
    python-lint-broken                           tracked
    python-lint-test                             tracked
    remove-python-shebangs                       tracked
    remove-python-shebangs-testing               tracked
    rtsp                                         tracked
  Local branches configured for 'git pull':
    new_shebangs           merges with remote remove-python-shebangs
    remove-python-shebangs merges with remote remove-python-shebangs
  Local refs configured for 'git push':
    Nachtzuster_improvements_to_birdnet_analysis pushes to Nachtzuster_improvements_to_birdnet_analysis (up to date)
    main                                         pushes to main                                         (fast-forwardable)
    notifications                                pushes to notifications                                (fast-forwardable)
    remove-python-shebangs                       pushes to remove-python-shebangs                       (up to date)
    rtsp                                         pushes to rtsp                                         (up to date)
+ sudo -u birder git reset --hard
HEAD is now at a786e51 Updating update_birdnet.sh to make more robust and offer functionality to switch branches for testing remote dev branches
+ set +x
+ sudo -u birder git fetch dev remove-python-shebangs
From https://github.com/jmherbst/BirdNET-Pi
 * branch            remove-python-shebangs -> FETCH_HEAD
+ set +x
+ sudo -u birder git switch -C remove-python-shebangs --track dev/remove-python-shebangs
Branch 'remove-python-shebangs' set up to track remote branch 'remove-python-shebangs' from 'dev'.
Switched to and reset branch 'remove-python-shebangs'
Your branch is up to date with 'dev/remove-python-shebangs'.
+ set +x
ExecStart=/home/birder/BirdNET-Pi/birdnet/bin/python3 /usr/local/bin/server.py
ExecStart=/home/birder/BirdNET-Pi/birdnet/bin/python3 /usr/local/bin/daily_plot.py
```

With that weird `test` failing based on where `stderr` is directed is odd to me, but also the only issue I've encountered. I'm going to do the PR to see about merging `ui-tweaks` into this branch on your fork, and then will try to run an update and see what we get.

Thank you so much for your help and these great contributions :+1:

File: scripts/update_birdnet.sh [hunk 2]
```diff
@@ -15,7 +15,7 @@ while getopts ":r:b:" o; do
       remote=${OPTARG}
 
       # Ensure that git remote is configured
-      git remote show $remote 2&> /dev/null || {
+      git remote show $remote 1&> /dev/null || {
```
Comment (mcguirepr89) [inline scripts/update_birdnet.sh:18]: I see now -- no `test` is actually envoked, which is why you have to retain the stderr for the `||` logic to work.

If you put it in a test I think you can dispense with all output
```
if git remote show $remote &>/dev/null;then
  remote=${OPTARG}
else
  echo "Error: remote '$remote' not found. Add the upstream remote to your repository and try again."
  exit 1
fi
```

**Comment by jmherbst** [comment]
@mcguirepr89 anything you need from me here? Just making sure you're not waiting on me for anything.

**Comment by mcguirepr89** [comment]
Hey thanks for checking -- I actually am waiting on you to accept the PR I submitted to this fork that would merge `ui-tweaks` so that I can test the update.

**Comment by jmherbst** [comment]
Hah, somehow I totally missed that - Thanks! I'll take a look

**Comment by jmherbst** [comment]
I don't want to muddy up this with a bunch of python lint fixes, so feel free to test/merge if you're alright with how tests turn out 👍

**Comment by mcguirepr89** [comment]
The update was successful -- I have a few additional things that I'll be doing to this, then one last merge of any last-minute changes to `ui-tweaks`, then the big merger along with a disclaimer that the changes are quite substantial and may require two updates or manually running a few `git` commands in case of issues.

PS -- take a look at all the cool new tweaks that you just pulled in from the `ui-tweaks` branch and definitely get a Flickr API key to enable the bird images throughout (thanks to @ehpersonal38 )

File: scripts/stats.php [hunk 3]
```diff
@@ -136,51 +170,52 @@
...
   flush();
-  $imagelink = "https://commons.wikimedia.org/w/index.php?search=$linkname&title=Special:MediaSearch&go=Go&type=image";
-  $homepage = file_get_contents($imagelink);
-  preg_match_all("{<img\\s*(.*?)src=('.*?'|\".*?\"|[^\\s]+)(.*?)\\s*/?>}ims", $homepage, $matches, PREG_SET_ORDER);
-  foreach ($matches as $val) {
-      $pos = strpos($val[2],"/");
-      $link = substr($val[2],1,-1);
-      if($pos !== 1 && strpos($link, "upload") == true && strpos($link, "CentralAutoLogin") == false)
-          echo "<img src=\"$link\">";
+  $flickrjson = json_decode(file_get_contents("https://www.flickr.com/services/rest/?method=flickr.photos.search&api_key=2158f6c5d66e89778bd0d340953f1bcf&text=\"".str_replace('_', '+', $comname)."\"&license=2%2C3%2C4%2C5%2C6%2C9&sort=relevance&per_page=15&format=json&nojsoncallback=1"), true)["photos"]["photo"];
```
Comment (jmherbst) [inline scripts/stats.php:179]: Hmm, was this someone's APi key leaked+hardcoded here? @mcguirepr89 + @ehpersonal38

Comment (mcguirepr89) [reply scripts/stats.php:179]: Yes!!! I imagine that belongs to @ehpersonal38 I would assume that key is meant to be `$config['FLIKR_API_KEY']` or something along those lines.

I should be able to clean it up

File: scripts/update_birdnet.sh [hunk 4]
```diff
@@ -12,13 +12,12 @@ branch="main"
 while getopts ":r:b:" o; do
   case "${o}" in
     r)
-      remote=${OPTARG}
-
-      # Ensure that git remote is configured
-      git remote show $remote > /dev/null 2>&1 || {
+      if git remote show $remote &>/dev/null;then
```
Comment (jmherbst) [inline scripts/update_birdnet.sh:15]: 2 issues with this I'm seeing:
1. We need to set remote to argument first before this conditional checks for it in configured remotes
2. It's not working for me 
```bash
$ ./scripts/update_birdnet.sh -r bogus_remote_name -b remove-python-shebangs
+ sudo -u birder git reset --hard
HEAD is now at 24e20fb This needs a proper test to work
+ set +x
+ sudo -u birder git fetch bogus_remote_name remove-python-shebangs
fatal: 'bogus_remote_name' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
+ set +x
+ sudo -u birder git switch -C remove-python-shebangs --track bogus_remote_name/remove-python-shebangs
fatal: invalid reference: bogus_remote_name/remove-python-shebangs
+ set +x
ExecStart=/home/birder/BirdNET-Pi/birdnet/bin/python3 /usr/local/bin/server.py
ExecStart=/home/birder/BirdNET-Pi/birdnet/bin/python3 /usr/local/bin/daily_plot.py
```

Comment (jmherbst) [reply scripts/update_birdnet.sh]: Please give [7a58b1c](https://github.com/mcguirepr89/BirdNET-Pi/pull/288/commits/7a58b1cc54e79607be7a0a653c37108c0833d22b) a test and let me know if this approach works for you.

Comment (mcguirepr89) [reply scripts/update_birdnet.sh]: Yes that worked thank you for spotting that obvious mistake! of course it needs to be set before the check is meaningful :)

**Comment by jmherbst** [comment]
We waiting on more testing here or more ui-tweaks?

**Comment by mcguirepr89** [comment]
Howdy!

I've thought about this and I think given the new method for updates in this PR, we can re-introduce the full python linting (instead of per line). What do you think?

**Comment by jmherbst** [comment]
Hey!  What do you think about following up this with a PR to roll out the linting changes?

**Comment by mcguirepr89** [comment]
That's probably a better idea! :+1:

**Comment by mcguirepr89** [comment]
This looks good now, too -- I'm going to merge this now and will try and update and will see what happens. :crossed_fingers:

**Comment by mcguirepr89** [comment]
Update went just fine!!! Thank you again for all of your work and patience!

**Pull request merged**